rfcs/convergence/root-as-slot.md

```diff
@@ -0,0 +1,241 @@
+# RFC: root as a slot
+
+---
+
+_@bsunderhus_
+
+## Summary
+
+This RFC proposes to treat `root` as a _slot_ to simplify signature, improve `Typings` and avoid conflicts between `root` and other _slots_ signatures.
+
+Main changes would be:
+
+1. `root` becomes a regular slot on component state
+2. Support references on every _slot_
+3. Supports Custom Elements on every _slot_
+4. Allows a restricted `as` prop on every slot
+5. stops using `getNativeElementProps` on every slot
+
+## Problem statement
+
+There are disparities in how `root` is treated compared to _slots_. Some discussions have popped up due to these divergences between _slots_ and `root`:
+
+1. [Do not support as prop for slots that render native DOM elements] conflicts with [1st rule of ARIA opt-out mechanism]
+2. Support for Custom Elements on _slots_ but not for _root_
+3. Right now `Typings` for `root` aren't available as there's no way to split `root` from internal state.
+4. There's no possible way of passing refs to slots but `root` does
+5. [Primary slot]
+
+Major differences between `root` 🌿 and _slots_ 🎰:
+
+1. References 🌿
+2. Custom Elements as base element 🎰
+3. `as` prop 🌿
+4. Well defined Types 🎰
+
+## Detailed Design or Proposal
+
+For the User of the component perspective there will be no changes in the signature.
+
+### Converge `root` to a _slot_
+
+This RFC proposes to treat `root` as a _slot_.
+
+Treating `root` as a slot means having to declare `root` _Typings_ together with other _slots_. This will impact in a components state signature, where `root` will have to be declared.
+
+> ⚠️ That doesn't mean `root` will be available in component's properties interface
+
+What we have right now:
+
+```ts
+export type Slots = {
+  slot: ObjectShorthandProps<SlotProps>;
+};
+export interface Props extends ComponentProps<Slots> & React.HTMLAttributes<HTMLElement> {/* ... */}
+export interface State extends ComponentState<Slots> {/* ... */}
+
+// use*State hook
+function useState({ slot, ...props }: Props): State {
+  // ...
+  return {
+    ...props,
+    slot: resolveShorthand(slot),
+  };
+}
+```
+
+With the modifications:
+
+```ts
+export type Slots = {
+  root: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>, HTMLElement>;
+  slot: ObjectShorthandProps<SlotProps>;
+};
+export interface Props extends ComponentProps<Slots> {
+  /* ... */
+}
+export interface State extends ComponentState<Slots> {
+  /* ... */
+}
+
+// use*State hook
+function useState({slot, ...props}: Props): State {
+  // ...
+  return {
+    slot: resolveShorthand(slot)
+    root: resolveShorthand(props),
+  };
+}
+```
+
+### Make _slots_ support `as` property (but with restrictions)
+
+Cancels [Do not support as prop for slots that render native DOM elements] as it already conflicts with [1st rule of ARIA opt-out mechanism]. This would add `as` property for `ObjectShorthandProps` signature for all _slots_.
```

Comment:
we haven't had an RFC yet that undoes the result of a previous one... I wonder if we should create a new `discarded` folder for that

Reply:
It's even trickier than that, because it's not undoing the whole RFC, is undoing one of the alterations proposed by the RFC heheheh

Reply:
Would amending the original RFC and keeping one source of truth make sense? Unsure what the best way forward here is.

Reply:
Honestly, I kind of have this feeling that we're going to end up with a sea of RFC's and eventually some of them will end up conflicting somewhere.

I'd propose go for the idea of having RFC's as a temporary thing and then having a more robust documentation that can be changed based on the proposals that come from the RFCs

Reply:
This is definitely something to be discussed!

Reply:
Also thought about that. RFCs could be considered one-shot proposals only and we should port that into doc friendly format, with only how the user should do X